File: hackerrank/problem solving/algorithms/2 - implementation/29 - find digits/find_digits.rb
# https://www.hackerrank.com/challenges/find-digits/problem?isFullScreen=true

def main
  n = gets.to_i
  n.times do
    num = gets.to_i
    puts find_digits(num)
  end
end

# k: quantity of digits of num = log num
# 0 < num < 10^9
# T: O(k) = O(9) = O(1)
# S: O(k) = O(9) = O(1) extra space
def find_digits(num)
  num.to_s.each_char.map(&:to_i).count { |d| !d.zero? && (num % d).zero? }
end

main if __FILE__ == $0

File: hackerrank/problem solving/algorithms/2 - implementation/29 - find digits/with morfo library/find_digits.java
// https://www.hackerrank.com/challenges/find-digits/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;

void main() {
    int n = Reader.read(Integer::parseInt);
    for (int i = 0; i < n; i++)
        IO.println(findDigits(Reader.read(Integer::parseInt)));
}

// k: quantity of digits of num = log num
// 0 < num < 10^9
// T: O(k) = O(9) = O(1)
// S: O(k) = O(9) = O(1) extra space
int findDigits(int num) {
    return (int) String.valueOf(num).chars().map(c -> c - '0').filter(d -> d != 0 && num % d == 0).count();
}
